# Dataframes Exercises

## Exercises Part I

#### 1. Run `python -m pip install pymysql` from your terminal to install the mysql client (any folder is fine)
#### 2. cd into your exercises folder for this module and run `echo env.py >> .gitignore`

#### 3. Create a function named `get_db_url`. It should accept a username, hostname, password, and database name and return a url connection string formatted like in the example at the start of this lesson.

In [1]:
import pandas as pd
import numpy as np

def get_db_url(db_name):
    from env import user, host, password
    return f'mysql+pymysql://{user}:{password}@{host}/{db_name}'

#### 4. Use your function to obtain a connection to the `employees` database.

In [2]:
sql = """
    SELECT * from dept_emp limit 3
"""
url = get_db_url("employees")

df = pd.read_sql(sql, url)
df

emp_no dept_no   from_date     to_date
0   10001    d005  1986-06-26  9999-01-01
1   10002    d007  1996-08-03  9999-01-01
2   10003    d004  1995-12-03  9999-01-01

#### 5. Once you have successfully run a query:

#### a. Intentionally make a typo in the database url. What kind of error message do you see?

In [3]:
pd.read_sql(sql, get_db_url('empolyees'))
# returns operational error

OperationalError: (pymysql.err.OperationalError) (1044, "Access denied for user 'hopper_1551'@'%' to database 'empolyees'")
(Background on this error at: http://sqlalche.me/e/14/e3q8)

#### b. Intentionally make an error in your SQL query. What does the error message look like?

In [4]:
sql = '''
    SELECT * FORM employees
    '''
pd.read_sql(sql, url)
# returns programming error stating sql syntax is bad, shows inputted sql

ProgrammingError: (pymysql.err.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'FORM employees' at line 1")
[SQL: 
    SELECT * FORM employees
    ]
(Background on this error at: http://sqlalche.me/e/14/f405)

#### 6. Read the `employees` and `titles` tables into two separate DataFrames.

In [5]:
sql = '''
    SELECT * FROM employees'''
url = get_db_url('employees')
employees_df = pd.read_sql(sql, url)
# above is same as this:
# employees_df = pd.read_sql('SELECT * FROM employees.employees',get_db_url('employees'))
employees_df

emp_no  birth_date first_name last_name gender   hire_date
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01
4        10005  1955-01-21    Kyoichi  Maliniak      M  1989-09-12
...        ...         ...        ...       ...    ...         ...
300019  499995  1958-09-24     Dekang  Lichtner      F  1993-01-12
300020  499996  1953-03-07       Zito      Baaz      M  1990-09-27
300021  499997  1961-08-03    Berhard    Lenart      M  1986-04-21
300022  499998  1956-09-05   Patricia   Breugel      M  1993-10-13
300023  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30

[300024 rows x 6 columns]

In [6]:
titles_df = pd.read_sql('SELECT * FROM titles', get_db_url('employees'))
titles_df

emp_no            title   from_date     to_date
0        10001  Senior Engineer  1986-06-26  9999-01-01
1        10002            Staff  1996-08-03  9999-01-01
2        10003  Senior Engineer  1995-12-03  9999-01-01
3        10004         Engineer  1986-12-01  1995-12-01
4        10004  Senior Engineer  1995-12-01  9999-01-01
...        ...              ...         ...         ...
443303  499997         Engineer  1987-08-30  1992-08-29
443304  499997  Senior Engineer  1992-08-29  9999-01-01
443305  499998     Senior Staff  1998-12-27  9999-01-01
443306  499998            Staff  1993-12-27  1998-12-27
443307  499999         Engineer  1997-11-30  9999-01-01

[443308 rows x 4 columns]

#### 7. How many rows and columns do you have in each DataFrame? Is that what you expected?

In [7]:
employees_df.shape

(300024, 6)

In [8]:
titles_df.shape

(443308, 4)

employees_df has 300024 rows × 6 columns

titles_df has 443308 rows × 4 columns

Yes, I did expect there to be more rows in the titles dataframe than in the employees dataframe as individual employees may have held multiple titles.

#### 8. Display the summary statistics for each DataFrame.

In [9]:
employees_df.describe()

emp_no
count  300024.000000
mean   253321.763392
std    161828.235540
min     10001.000000
25%     85006.750000
50%    249987.500000
75%    424993.250000
max    499999.000000

In [10]:
employees_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 300024 entries, 0 to 300023
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   emp_no      300024 non-null  int64 
 1   birth_date  300024 non-null  object
 2   first_name  300024 non-null  object
 3   last_name   300024 non-null  object
 4   gender      300024 non-null  object
 5   hire_date   300024 non-null  object
dtypes: int64(1), object(5)
memory usage: 13.7+ MB


In [11]:
titles_df.describe()

emp_no
count  443308.000000
mean   253075.034430
std    161853.292613
min     10001.000000
25%     84855.750000
50%    249847.500000
75%    424891.250000
max    499999.000000

In [12]:
titles_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 443308 entries, 0 to 443307
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype 
---  ------     --------------   ----- 
 0   emp_no     443308 non-null  int64 
 1   title      443308 non-null  object
 2   from_date  443308 non-null  object
 3   to_date    443308 non-null  object
dtypes: int64(1), object(3)
memory usage: 13.5+ MB


#### 9. How many unique titles are in the `titles` DataFrame?

In [13]:
pd.read_sql('SELECT COUNT(DISTINCT title) FROM titles', url)
# 7 unique titles

COUNT(DISTINCT title)
0                      7

#### 10. What is the oldest date in the `to_date` column?

In [14]:
sql10 = '''SELECT MIN(to_date)
FROM titles;'''
pd.read_sql(sql10,url)
# 1985-03-01

MIN(to_date)
0   1985-03-01

#### 11. What is the most recent date in the `to_date` column?

In [15]:
# I interpreted 'most recent' as meaning the latest date that is not in the future
sql11 = '''
SELECT MAX(to_date)
FROM titles
WHERE to_date <= CURDATE();'''
pd.read_sql(sql11,url)
# 2002-08-01

MAX(to_date)
0   2002-08-01

In [16]:
# to change dtype to date
titles_df.from_date = pd.to_datetime(titles_df.from_date)
titles_df.dtypes

emp_no                int64
title                object
from_date    datetime64[ns]
to_date              object
dtype: object

## Exercises Part II

#### 1. Copy the `users` and `roles` DataFrames from the examples above.

In [286]:
users = pd.DataFrame({
    'id': [1, 2, 3, 4, 5, 6],
    'name': ['bob', 'joe', 'sally', 'adam', 'jane', 'mike'],
    'role_id': [1, 2, 3, 3, np.nan, np.nan]
})
users

id   name  role_id
0   1    bob      1.0
1   2    joe      2.0
2   3  sally      3.0
3   4   adam      3.0
4   5   jane      NaN
5   6   mike      NaN

In [18]:
roles = pd.DataFrame({
    'id': [1, 2, 3, 4],
    'name': ['admin', 'author', 'reviewer', 'commenter']
})
roles

id       name
0   1      admin
1   2     author
2   3   reviewer
3   4  commenter

#### 2. What is the result of using a `right` join on the DataFrames?

In [69]:
users.merge(roles, how='right', left_on = 'role_id', right_on = 'id')

id_x name_x  role_id  id_y     name_y
0   1.0    bob      1.0     1      admin
1   2.0    joe      2.0     2     author
2   3.0  sally      3.0     3   reviewer
3   4.0   adam      3.0     3   reviewer
4   NaN    NaN      NaN     4  commenter

#### 3. What is the result of using an `outer` join on the DataFrames?

In [70]:
users.merge(roles, how='outer', left_on = 'role_id', right_on = 'id')

id_x name_x  role_id  id_y     name_y
0   1.0    bob      1.0   1.0      admin
1   2.0    joe      2.0   2.0     author
2   3.0  sally      3.0   3.0   reviewer
3   4.0   adam      3.0   3.0   reviewer
4   5.0   jane      NaN   NaN        NaN
5   6.0   mike      NaN   NaN        NaN
6   NaN    NaN      NaN   4.0  commenter

#### 4. What happens if you drop the foreign keys from the DataFrames and try to merge them?

In [294]:
#users.drop(columns='role_id', inplace=True)
#users.drop(columns='id', inplace=True)
#roles.drop(columns='id', inplace=True)
users.merge(roles, how='outer')
# cannot merge them in a meaningful way

name
0        bob
1        joe
2      sally
3       adam
4       jane
5       mike
6      admin
7     author
8   reviewer
9  commenter

#### 5. Load the `mpg` dataset from PyDataset.

In [28]:
from pydataset import data

In [29]:
mpg = data('mpg')
mpg

manufacturer   model  displ  year  cyl       trans drv  cty  hwy fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f   18   29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f   21   29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f   20   31  p   
4           audi      a4    2.0  2008    4    auto(av)   f   21   30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f   16   26  p   
..           ...     ...    ...   ...  ...         ...  ..  ...  ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f   19   28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f   21   29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f   16   26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f   18   26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f   17   26  p   

       class  
1    compact  
2    compact  
3    compact  
4    compact  
5    compact  
..       ...  
230  midsize  
231  midsize  
232  midsize  
233  midsize  
234  midsize  

[234 rows x 11 columns]

#### 6. Output and read the documentation for the `mpg` dataset.

In [30]:
data('mpg', show_doc=True)

mpg

PyDataset Documentation (adopted from R Documentation. The displayed examples are in R)

## Fuel economy data from 1999 and 2008 for 38 popular models of car

### Description

This dataset contains a subset of the fuel economy data that the EPA makes
available on http://fueleconomy.gov. It contains only models which had a new
release every year between 1999 and 2008 - this was used as a proxy for the
popularity of the car.

### Usage

    data(mpg)

### Format

A data frame with 234 rows and 11 variables

### Details

  * manufacturer. 

  * model. 

  * displ. engine displacement, in litres 

  * year. 

  * cyl. number of cylinders 

  * trans. type of transmission 

  * drv. f = front-wheel drive, r = rear wheel drive, 4 = 4wd 

  * cty. city miles per gallon 

  * hwy. highway miles per gallon 

  * fl. 

  * class. 




#### 7. How many rows and columns are in the dataset?

In [31]:
mpg.shape
# 234 rows x 11 columns

(234, 11)

#### 8. Check out your column names and perform any cleanup you may want on them.

In [32]:
mpg.columns

Index(['manufacturer', 'model', 'displ', 'year', 'cyl', 'trans', 'drv', 'cty',
       'hwy', 'fl', 'class'],
      dtype='object')

In [40]:
mpg.rename(columns={'cty':'city', 'hwy':'highway'}, inplace=True)

#### 9. Display the summary statistics for the dataset.

In [41]:
mpg.describe()

displ         year         cyl        city     highway
count  234.000000   234.000000  234.000000  234.000000  234.000000
mean     3.471795  2003.500000    5.888889   16.858974   23.440171
std      1.291959     4.509646    1.611534    4.255946    5.954643
min      1.600000  1999.000000    4.000000    9.000000   12.000000
25%      2.400000  1999.000000    4.000000   14.000000   18.000000
50%      3.300000  2003.500000    6.000000   17.000000   24.000000
75%      4.600000  2008.000000    8.000000   19.000000   27.000000
max      7.000000  2008.000000    8.000000   35.000000   44.000000

#### 10. How many different manufacturers are there?

In [77]:
mpg.manufacturer.nunique()
# 15

15

#### 11. How many different models are there?

In [36]:
mpg.model.nunique()
# 38

38

#### 12. Create a column named `mileage_difference` like you did in the DataFrames exercises; this column should contain the difference between highway and city mileage for each car.

In [44]:
mpg['mileage_difference'] = mpg.highway-mpg.city
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  
1    compact                  11  
2    compact                   8  
3    compact                  11  
4    compact                   9  
5    compact                  10  
..       ...                 ...  
230  midsize                   9  
231  midsize                   8  
232  midsize                  10  
233  midsize                   8  
234  midsize                   9  

[234 rows x 12 columns]

#### 13. Create a column named `average_mileage` like you did in the DataFrames exercises; this is the mean of the city and highway mileage.

In [45]:
mpg['average_mileage'] = (mpg.city + mpg.highway)/2
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  average_mileage  
1    compact                  11             23.5  
2    compact                   8             25.0  
3    compact                  11             25.5  
4    compact                   9             25.5  
5    compact                  10             21.0  
..       ...                 ...              ...  
230  midsize                   9             23.5  
231  midsize                   8             25.0  
232  midsize                  10             21.0  
233  midsize                   8             22.0  
234  midsize                   9             21.5  

[234 rows x 13 columns]

#### 14. Create a new column on the `mpg` dataset named `is_automatic` that holds boolean values denoting whether the car has an automatic transmission.

In [66]:
mpg['is_automatic'] = mpg.trans.str.contains('auto')
mpg

manufacturer   model  displ  year  cyl       trans drv  city  highway fl  \
1           audi      a4    1.8  1999    4    auto(l5)   f    18       29  p   
2           audi      a4    1.8  1999    4  manual(m5)   f    21       29  p   
3           audi      a4    2.0  2008    4  manual(m6)   f    20       31  p   
4           audi      a4    2.0  2008    4    auto(av)   f    21       30  p   
5           audi      a4    2.8  1999    6    auto(l5)   f    16       26  p   
..           ...     ...    ...   ...  ...         ...  ..   ...      ... ..   
230   volkswagen  passat    2.0  2008    4    auto(s6)   f    19       28  p   
231   volkswagen  passat    2.0  2008    4  manual(m6)   f    21       29  p   
232   volkswagen  passat    2.8  1999    6    auto(l5)   f    16       26  p   
233   volkswagen  passat    2.8  1999    6  manual(m5)   f    18       26  p   
234   volkswagen  passat    3.6  2008    6    auto(s6)   f    17       26  p   

       class  mileage_difference  average_mileage  is_automatic  
1    compact                  11             23.5          True  
2    compact                   8             25.0         False  
3    compact                  11             25.5         False  
4    compact                   9             25.5          True  
5    compact                  10             21.0          True  
..       ...                 ...              ...           ...  
230  midsize                   9             23.5          True  
231  midsize                   8             25.0         False  
232  midsize                  10             21.0          True  
233  midsize                   8             22.0         False  
234  midsize                   9             21.5          True  

[234 rows x 14 columns]

#### 15. Using the `mpg` dataset, find out which which manufacturer has the best miles per gallon on average?

In [62]:
mpg.groupby('manufacturer').average_mileage.mean().sort_values(ascending=False).head(1)
# honda

manufacturer
honda    28.5
Name: average_mileage, dtype: float64

#### 16. Do automatic or manual cars have better miles per gallon?

In [67]:
mpg.groupby('is_automatic').average_mileage.mean()
# manual cars

is_automatic
False    22.227273
True     19.130573
Name: average_mileage, dtype: float64

## Exercises Part III

#### 1. Use your `get_db_url` function to help you explore the data from the `chipotle` database.

In [125]:
url = get_db_url('chipotle')
sql = 'SHOW TABLES'
pd.read_sql(sql, url)

Tables_in_chipotle
0             orders

In [143]:
sql = 'SELECT * FROM orders'
chipotle = pd.read_sql(sql,url)
chipotle

id  order_id  quantity                              item_name  \
0        1         1         1           Chips and Fresh Tomato Salsa   
1        2         1         1                                   Izze   
2        3         1         1                       Nantucket Nectar   
3        4         1         1  Chips and Tomatillo-Green Chili Salsa   
4        5         2         2                           Chicken Bowl   
...    ...       ...       ...                                    ...   
4617  4618      1833         1                          Steak Burrito   
4618  4619      1833         1                          Steak Burrito   
4619  4620      1834         1                     Chicken Salad Bowl   
4620  4621      1834         1                     Chicken Salad Bowl   
4621  4622      1834         1                     Chicken Salad Bowl   

                                                                     choice_description  \
0                                                                                   nan   
1                                                                          [Clementine]   
2                                                                               [Apple]   
3                                                                                   nan   
4            [Tomatillo-Red Chili Salsa (Hot), [Black Beans, Rice, Cheese, Sour Cream]]   
...                                                                                 ...   
4617  [Fresh Tomato Salsa, [Rice, Black Beans, Sour Cream, Cheese, Lettuce, Guacamole]]   
4618               [Fresh Tomato Salsa, [Rice, Sour Cream, Cheese, Lettuce, Guacamole]]   
4619         [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Guacamole, Lettuce]]   
4620                                 [Fresh Tomato Salsa, [Fajita Vegetables, Lettuce]]   
4621                    [Fresh Tomato Salsa, [Fajita Vegetables, Pinto Beans, Lettuce]]   

     item_price  
0        $2.39   
1        $3.39   
2        $3.39   
3        $2.39   
4       $16.98   
...         ...  
4617    $11.75   
4618    $11.75   
4619    $11.25   
4620     $8.75   
4621     $8.75   

[4622 rows x 6 columns]

In [127]:
chipotle.dtypes

id                     int64
order_id               int64
quantity               int64
item_name             object
choice_description    object
item_price            object
dtype: object

#### 2. What is the total price for each order?

In [145]:
chipotle['item_price'] = chipotle.item_price.str.replace('$','').astype(float)
chipotle.dtypes

<ipython-input-145-75d9a54cf7a6>:1: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  chipotle['item_price'] = chipotle.item_price.str.replace('$','').astype(float)


id                      int64
order_id                int64
quantity                int64
item_name              object
choice_description     object
item_price            float64
dtype: object

In [146]:
chipotle.pivot_table(aggfunc='sum', index='order_id', columns='item_name', values='item_price', margins=True)['All']

order_id
1          11.56
2          16.98
3          12.67
4          21.00
5          13.70
          ...   
1831       12.90
1832       13.20
1833       23.50
1834       28.75
All     34500.16
Name: All, Length: 1835, dtype: float64

#### 3. What are the most popular 3 items?

In [168]:
qty_ordered_per_item = chipotle.groupby('item_name').count()['quantity']
qty_ordered_per_item.sort_values(ascending=False).head(3)

item_name
Chicken Bowl           726
Chicken Burrito        553
Chips and Guacamole    479
Name: quantity, dtype: int64

#### 4. Which item has produced the most revenue?

In [316]:
chipotle[['item_name', 'item_price']].groupby('item_name').agg('sum').idxmax()

item_price    Chicken Bowl
dtype: object

#### 5. Join the `employees` and `titles` DataFrames together.

In [319]:
emp_titles_df = employees_df.merge(titles_df, left_on='emp_no', right_on='emp_no', how='inner')
emp_titles_df

emp_no  birth_date first_name last_name gender   hire_date  \
0        10001  1953-09-02     Georgi   Facello      M  1986-06-26   
1        10002  1964-06-02    Bezalel    Simmel      F  1985-11-21   
2        10003  1959-12-03      Parto   Bamford      M  1986-08-28   
3        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
4        10004  1954-05-01  Chirstian   Koblick      M  1986-12-01   
...        ...         ...        ...       ...    ...         ...   
443303  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443304  499997  1961-08-03    Berhard    Lenart      M  1986-04-21   
443305  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443306  499998  1956-09-05   Patricia   Breugel      M  1993-10-13   
443307  499999  1958-05-01     Sachin   Tsukuda      M  1997-11-30   

                  title  from_date     to_date  
0       Senior Engineer 1986-06-26  9999-01-01  
1                 Staff 1996-08-03  9999-01-01  
2       Senior Engineer 1995-12-03  9999-01-01  
3              Engineer 1986-12-01  1995-12-01  
4       Senior Engineer 1995-12-01  9999-01-01  
...                 ...        ...         ...  
443303         Engineer 1987-08-30  1992-08-29  
443304  Senior Engineer 1992-08-29  9999-01-01  
443305     Senior Staff 1998-12-27  9999-01-01  
443306            Staff 1993-12-27  1998-12-27  
443307         Engineer 1997-11-30  9999-01-01  

[443308 rows x 9 columns]

#### 6. For each title, find the hire date of the employee that was hired most recently with that title.

In [275]:
emp_titles_df.hire_date = pd.to_datetime(emp_titles_df.hire_date)
emp_titles_df.dtypes

emp_no                 int64
birth_date            object
first_name            object
last_name             object
gender                object
hire_date     datetime64[ns]
title                 object
from_date     datetime64[ns]
to_date               object
dtype: object

In [320]:
emp_titles_df.pivot_table('hire_date', 'title', aggfunc='max')

hire_date
title                         
Assistant Engineer  1999-12-24
Engineer            2000-01-28
Manager             1992-02-05
Senior Engineer     2000-01-01
Senior Staff        2000-01-13
Staff               2000-01-12
Technique Leader    1999-12-31

#### 7. Write the code necessary to create a cross tabulation of the number of titles by department. (Hint: this will involve a combination of SQL code to pull the necessary data and python/pandas code to perform the manipulations.)

In [207]:
sql = 'SELECT * FROM dept_emp'
dept_emp_df = pd.read_sql(sql,get_db_url("employees"))
dept_emp_df

emp_no dept_no   from_date     to_date
0        10001    d005  1986-06-26  9999-01-01
1        10002    d007  1996-08-03  9999-01-01
2        10003    d004  1995-12-03  9999-01-01
3        10004    d004  1986-12-01  9999-01-01
4        10005    d003  1989-09-12  9999-01-01
...        ...     ...         ...         ...
331598  499995    d004  1997-06-02  9999-01-01
331599  499996    d004  1996-05-13  9999-01-01
331600  499997    d005  1987-08-30  9999-01-01
331601  499998    d002  1993-12-27  9999-01-01
331602  499999    d004  1997-11-30  9999-01-01

[331603 rows x 4 columns]

In [251]:
sql = 'SELECT * FROM departments'
departments_df = pd.read_sql(sql,get_db_url("employees"))
departments_df

dept_no           dept_name
0    d009    Customer Service
1    d005         Development
2    d002             Finance
3    d003     Human Resources
4    d001           Marketing
5    d004          Production
6    d006  Quality Management
7    d008            Research
8    d007               Sales

In [323]:
emp_title_dept = emp_titles_df.merge(dept_emp_df, left_on='emp_no', right_on='emp_no')
emp_title_dept = emp_title_dept.drop(columns={'birth_date','gender'}).rename(columns={'from_date_x':'title_from', 'to_date_x':'title_until', 'from_date_y':'in_dept_from', 'to_date_y':'in_dept_until'})
emp_title_dept = emp_title_dept.merge(departments_df)
emp_title_dept
#emp_title_dept[(emp_title_dept['title_until'] == '9999-01-01') & (emp_title_dept['in_dept_until'] == '9999-01-01')]

emp_no first_name  last_name   hire_date               title  \
0        10001     Georgi    Facello  1986-06-26     Senior Engineer   
1        10006     Anneke    Preusig  1989-06-02     Senior Engineer   
2        10008     Saniya   Kalloufi  1994-09-15  Assistant Engineer   
3        10012   Patricio  Bridgland  1992-12-18            Engineer   
4        10012   Patricio  Bridgland  1992-12-18     Senior Engineer   
...        ...        ...        ...         ...                 ...   
489898  499975     Masali    Chorvat  1992-01-23               Staff   
489899  499977    Martial    Weisert  1996-09-17               Staff   
489900  499989  Keiichiro  Lindqvist  1993-10-28        Senior Staff   
489901  499998   Patricia    Breugel  1993-10-13        Senior Staff   
489902  499998   Patricia    Breugel  1993-10-13               Staff   

       title_from title_until dept_no in_dept_from in_dept_until    dept_name  
0      1986-06-26  9999-01-01    d005   1986-06-26    9999-01-01  Development  
1      1990-08-05  9999-01-01    d005   1990-08-05    9999-01-01  Development  
2      1998-03-11  2000-07-31    d005   1998-03-11    2000-07-31  Development  
3      1992-12-18  2000-12-18    d005   1992-12-18    9999-01-01  Development  
4      2000-12-18  9999-01-01    d005   1992-12-18    9999-01-01  Development  
...           ...         ...     ...          ...           ...          ...  
489898 1995-11-21  1997-07-23    d002   1996-10-19    1997-07-23      Finance  
489899 1999-12-28  9999-01-01    d002   1999-12-28    9999-01-01      Finance  
489900 1996-10-29  2001-03-07    d002   1996-10-29    2001-03-07      Finance  
489901 1998-12-27  9999-01-01    d002   1993-12-27    9999-01-01      Finance  
489902 1993-12-27  1998-12-27    d002   1993-12-27    9999-01-01      Finance  

[489903 rows x 11 columns]

In [312]:
v = pd.crosstab(emp_title_dept.title, emp_title_dept.dept_name)#.apply(lambda value: np.where(value > 0, 1, 0))
v

dept_name           Customer Service  Development  Finance  Human Resources  \
title                                                                         
Assistant Engineer               298         7769        0                0   
Engineer                        2362        58135        0                0   
Manager                            4            2        2                2   
Senior Engineer                 2027        49326        0                0   
Senior Staff                   13925         1247    12139            12274   
Staff                          16150         1424    13929            14342   
Technique Leader                 309         7683        0                0   

dept_name           Marketing  Production  Quality Management  Research  Sales  
title                                                                           
Assistant Engineer          0        6445                1831       378      0  
Engineer                    0       49649               13852      2986      0  
Manager                     2           4                   4         2      2  
Senior Engineer             0       42205               11864      2570      0  
Senior Staff            13940        1270                   0     11637  36191  
Staff                   16196        1478                   0     13495  41808  
Technique Leader            0        6557                1795       393      0